### PR TITLE
Add Playlist Features

### PR DESCRIPTION
**Pull request by Dumeng**
- [x] Move storage from LocalStorage to indexedDB
- [x] <del>Add Modal</del>
- [x] Add Online Playlist Bookmark
- [x] <del>Add Local Playlist Creation</del>
- [x] Existing Data Migration
- [x] Add Sidebar Playlist List
- [x] Add Event Handler for Playlist Update

### CODE REVIEW

**Comment by listen1** [comment]
indexdb is a nice optimaze way to manage huge playlist. nice pr.
i'm also curious about what are those unfinished task about? 😀 
maybe we can divide them into seperated pr.

**Comment by Dumeng** [comment]
Yes, indexeddb is used for playlist persistence. So I want to fully implement it before merging. There are some details needs to be designed, so it will take some time.
The modal dialogue is required for playlist creation. But I can migrate existing playlist first.

**Comment by mikelxk** [comment]
@Dumeng Would you mind rebase #673 which contains the refactor in player from vuex to composition api into this pr.

**Comment by Dumeng** [comment]
> @Dumeng Would you mind merge #673 which contains the refactor in player from vuex to composition api into this pr.

I will resolve the code merge conflict on this branch after your branch merged.

**Comment by Dumeng** [comment]
@listen1
I will create another PR for modal and local playlist. This PR is ready for review now.

**Comment by listen1** [comment]
> @listen1 I will create another PR for modal and local playlist. This PR is ready for review now.

sure. I'm doing review and testing related features.

**Comment by listen1** [comment]
localstorage data includes two different types of data: 
* fixed key name
  * theme
  * language
  * enableXXX
  * player-settings
  * current-playing: now playing tracks
  * playerlists: myplaylist ids
  * favoriteplayerlists: favorite playlist ids
* flexible key name:
  * [playlist_id]: store playlist info and tracks included.

new dexie db use new tables to store these data:
* settings: fixed key name data except playlists and favorateplaylists key data
* playlists: playlists and favorateplaylists data except tracks included
* tracks: flexible key name

design is nice for efficient query and manage data.

Here're some suggestion for this pr to merge:

1. sequence
we store myplaylist ids in `playerlists` entry in localstorage, so playlist is kept in sequence. If these data store in playlist table, maybe a sequence field is needed to indicate playlist position in all playlist, so users can adjust it. I noticed in myplaylist.js file, related function still use `localStorage` api to save, which may cause some problem.

tracks in playlist also needs to keep sequence.

2. data migration code
data migration code seems only handle `current-playing` entry in history data. I'm not sure if I miss other part for migration.

**Comment by Dumeng** [comment]
Thanks for the reviewing and documentation.
1. The settings table is designed as a key-value table. We can add new key-value pairs to store the order of the playlists. If it's possible. I'd like to minimize the usage of localstorage in the new version to simplify the complexity and reduce confusion if it's possible.
2. The migration functionality heavily relay on implements of the original features. And it's may damage the old data. I will add the settings migration code. But the playlist migration may need implement the local playlist related features firstly, so we can verify if the migrated data work properly.

**Comment by listen1** [comment]
> Thanks for the reviewing and documentation.
> 
> 1. The settings table is designed as a key-value table. We can add new key-value pairs to store the order of the playlists. If it's possible. I'd like to minimize the usage of localstorage in the new version to simplify the complexity and reduce confusion if it's possible.

I think there're two ways to keep sequence data:
1. Store in settings as we do before
2. Store sequence number as column field in playlist table and track table

I think both solution will work. I prefer second way because if there're lots of tracks in playlist, read key-value field will cost a lot to read all the track id in playlist and then fetch first n elements. That will break our target (for improve query efficiency). If use sequece number field in table, we can use orderBy function to fetch data.

> 2. The migration functionality heavily relay on implements of the original features. And it's may damage the old data. I will add the settings migration code. But the playlist migration may need implement the local playlist related features firstly, so we can verify if the migrated data work properly.

Sure. Migration feature can be added later as another PR.

**Comment by Dumeng** [comment]
The option 1 can be optimized. We can only store ids into an array. This can store the order with much lower overhead. And it's also easy to handle re-order operations.

As for the option 2, it's a typical approach to store sorted list in to database. But it's also hard to design a good solution for the re-order operation which will easily cause a lot item update. Eg. move the last item to the first position in below table will update almost all items. 
| id | order |
|--|--|
| 1 | 1 |
| 2 | 2 |
| ... | ... |
| 100 | 100 |
| **101** | 101 |

V V V

| id | order |
|--|--|
| **101** | 1 |
| 1 | 2 |
| 2 | 3 |
| ... | ... |
| 100 | 101 |

**Comment by listen1** [comment]
> The option 1 can be optimized. We can only store ids into an array. This can store the order with much lower overhead. And it's also easy to handle re-order operations.
> 
> As for the option 2, it's a typical approach to store sorted list in to database. But it's also hard to design a good solution for the re-order operation which will easily cause a lot item update. Eg. move the last item to the first position in below table will update almost all items.
> 
> id	order
> 1	1
> 2	2
> ...	...
> 100	100
> **101**	101
> V V V
> 
> id	order
> **101**	1
> 1	2
> 2	3
> ...	...
> 100	101

Yes, it will cost a lot when user try to reorder a big list, update order field when we make modification. If we use option 1, we also need to "rewrite" whole id list. 

Update whole order field when reordering seems silly, but it's simple to understand and works fine in most situations. refer to https://stackoverflow.com/a/10356675

Situation A: query first page (page size 20) for tracks in a big playlist (including 10000 tracks)
option1: get track ids (10000 items), slice first 20 elements, do indexed db query
option2: do indexed by query using order by (order field, limit 20)

Situation B: reorder a big playlist (including 10000 tracks)
option1: rewrite track ids list (10000 ids value)
option2: update order field for n ( 1<n<10000) rows

Those 2 situation I can hardly say which option works better. In my opnion, using option 1 or option 2 are all acceptable. Since this pr is for optimaze localstorage performance, option 2 seems to use indexed db feature to accelerate query performance. 

btw: there're some way to optimaze option2, refer to https://stackoverflow.com/questions/5651299/store-ordered-list-in-database-gap-approach , I think it's too complex and hard to maintain.

**Comment by Dumeng** [comment]
Store order to array for now. The order field for each entry is a little hard to implement. But we have chance to change it before release.

**Comment by Dumeng** [comment]
Add a new package mitt as the global event bus. We can use it to communicate between components.

**Comment by Dumeng** [comment]
Hi @listen1 ,
The my playlist feature and new modal components are implemented. Shall I add them to this PR?
For the playlist order feature, please let me if further modification needed.

**Comment by listen1** [comment]
> Hi @listen1 , The my playlist feature and new modal components are implemented. Shall I add them to this PR? For the playlist order feature, please let me if further modification needed.

If new features are based on new indexeddb solution, you can add them to this pr.
by the way, I think order of tracks are almost done, playlist order feature is not implemented. (which means the sequence of my playlist or fav playlists are shown in sidebar)

**Comment by mikelxk** [comment]
Hi, @Dumeng , I saw you put `Sidebar.vue` in `src\components`. I believed components should refer to some code structure that can be reused. The sidebar now seem only to be a part of the page which is not reused. Is it better to move it to `src/views`? 🤔

**Comment by Dumeng** [comment]
> Hi, @Dumeng , I saw you put `Sidebar.vue` in `src\components`. I believed components should refer to some code structure that can be reused. The sidebar now seem only to be a part of the page which is not reused. Is it better to move it to `src/views`? 🤔

In my opinion, Views should be for routing purpose. Components is just parts for a view.

**Comment by mikelxk** [comment]
I did a fresh install of the fork and it says and when I try to add a playlist it shows:
`_ctx.clonePlaylist is not a function
    at createBaseVNode.onClick._cache.<computed>._cache.` 
And I also get:
`_ctx.closeWindow is not a function
    at createBaseVNode.onClick._cache.<computed>._cache.`
 when I pressed refresh button

**Comment by Dumeng** [comment]
@mikelxk  Thanks for your feedback, the missing button functions are added. Some other features are still under development.

**Comment by mikelxk** [comment]
> @mikelxk Thanks for your feedback, the missing button functions are added. Some other features are still under development.

Thanks for the quick fix, there's a few bugs I found:
1. When I favorite a list. The sidebar will completely disappear if I cancel favorite.
2. The i18n locale is null when it's first installed.

File: src/components/Modal.vue [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  DefaultModal,
+  CreatePlaylist,
+  AddToPlaylist,
+};
+
+const showModal = (type = 'DefaultModal', opt: Record<string, unknown> = {}) => {
+  show = true;
+  modalType = components[type] || DefaultModal;
+  options = opt;
+  console.log(type);
```
Comment (listen1) [inline src/components/Modal.vue:30]: please clean unused debug console before commit

File: src/components/modals/DefaultModal.vue [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          <slot name="footer">
+            default footer
+            <button class="modal-default-button" @click="$emit('close')">OK</button>
+          </slot>
+        </div>
+      </div>
+    </div>
+  </div>
+</template>
+
```
Comment (listen1) [inline src/components/modals/DefaultModal.vue:26]: bug found, click close button top right not working. 

we can add code below to fix.

```vue
<script setup lang="ts">
const emit = defineEmits(['close']);
</script>
```
reason is close event is not relay to parent when use `DefaultModal`

Comment (Dumeng) [reply src/components/modals/DefaultModal.vue:26]: Fix this with an alternative method.
`$parent.$emit('close')`

**Pull request merged**